### PR TITLE
feat: allow access approvals with user groups

### PR DESCRIPTION
**Pull request by meetcshah19**
# Description 📣
Allow secret access approvals using user groups.
Demo : https://github.com/user-attachments/assets/4f839d1f-8442-4da6-9163-a91866fbdba6

## Type ✨

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation

- [ ] I have read the [contributing guide](https://infisical.com/docs/contributing/getting-started/overview), agreed and acknowledged the [code of conduct](https://infisical.com/docs/contributing/getting-started/code-of-conduct). 📝

### CODE REVIEW

**Review by DanielHougaard — CHANGES_REQUESTED** [review]

File: backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts [hunk 1]
```diff
@@ -69,14 +80,32 @@ export const accessApprovalPolicyServiceFactory = ({
...
+        email: string | null | undefined;
+        username: string;
+        firstName: string | null | undefined;
+        lastName: string | null | undefined;
+        isPartOfGroup: boolean;
+      }[]
+    >[] = [];
+
+    for (const groupId of groupApprovers) {
+      usersPromises.push(groupDAL.findAllGroupMembers({ orgId: actorOrgId, groupId, offset: 0 }));
```
[the remark this answers is not in the document]
Comment (maidul98) [reply backend/src/ee/services/access-approval-policy/access-approval-policy-service.ts:96]: Might be best to keep as is for readability purposes

File: backend/src/ee/routes/v1/access-approval-policy-router.ts [hunk 2]
```diff
@@ -125,23 +119,20 @@ export const registerAccessApprovalPolicyRouter = async (server: FastifyZodProvi
...
+        name: z.string().optional(),
+        secretPath: z
+          .string()
+          .trim()
+          .optional()
+          .transform((val) => (val === "" ? "/" : val)),
+        approvers: z
+          .object({ type: z.nativeEnum(ApproverType), id: z.string() })
+          .array()
+          .min(1),
```
Comment (DanielHougaard) [inline backend/src/ee/routes/v1/access-approval-policy-router.ts:132]: Same as above

File: backend/src/ee/routes/v1/access-approval-policy-router.ts [hunk 3]
```diff
@@ -11,21 +12,18 @@ export const registerAccessApprovalPolicyRouter = async (server: FastifyZodProvi
...
-        }),
+      body: z.object({
+        projectSlug: z.string().trim(),
+        name: z.string().optional(),
+        secretPath: z.string().trim().default("/"),
+        environment: z.string(),
+        approvers: z
+          .object({ type: z.nativeEnum(ApproverType), id: z.string() })
+          .array()
+          .min(1),
```
Comment (DanielHougaard) [inline backend/src/ee/routes/v1/access-approval-policy-router.ts:23]: I would add a custom message on the approvers array.

I forget the exact syntax, but I think it's just a separate parameter like so: `.min(1, { message: "At least one approver should be provided" } )`

File: backend/src/ee/routes/v1/secret-approval-policy-router.ts [hunk 4]
```diff
@@ -16,26 +17,23 @@ export const registerSecretApprovalPolicyRouter = async (server: FastifyZodProvi
...
+        environment: z.string(),
+        secretPath: z
+          .string()
+          .optional()
+          .nullable()
+          .default("/")
+          .transform((val) => (val ? removeTrailingSlash(val) : val)),
+        approvers: z
+          .object({ type: z.nativeEnum(ApproverType), id: z.string() })
+          .array()
```
Comment (DanielHougaard) [inline backend/src/ee/routes/v1/secret-approval-policy-router.ts:32]: Same as above

File: backend/src/ee/routes/v1/secret-approval-policy-router.ts [hunk 5]
```diff
@@ -68,24 +66,21 @@ export const registerSecretApprovalPolicyRouter = async (server: FastifyZodProvi
...
-            .transform((val) => (val === "" ? "/" : val)),
-          enforcementLevel: z.nativeEnum(EnforcementLevel).optional()
-        })
-        .refine((data) => data.approvers.length > 0 || data.groupApprovers.length > 0, {
-          path: ["approvers", "groupApprovers"],
-          message: "At least one approver should be provided."
-        }),
+      body: z.object({
+        name: z.string().optional(),
+        approvers: z
```
Comment (DanielHougaard) [inline backend/src/ee/routes/v1/secret-approval-policy-router.ts:71]: Same as above

File: backend/src/ee/services/access-approval-policy/access-approval-policy-dal.ts [hunk 6]
```diff
@@ -56,16 +53,18 @@ export const accessApprovalPolicyDALFactory = (db: TDbClient) => {
...
           },
           {
             key: "approverGroupId",
-            label: "groupApprovers" as const,
-            mapper: ({ approverGroupId }) => ({
-              groupId: approverGroupId
+            label: "approvers" as const,
+            mapper: ({ approverGroupId: id }) => ({
+              id,
+              type: "group"
```
Comment (DanielHougaard) [inline backend/src/ee/services/access-approval-policy/access-approval-policy-dal.ts]: Please use the `ApproverType` enum here

File: backend/src/ee/services/secret-approval-request/secret-approval-request-service.ts [hunk 7]
```diff
@@ -447,8 +447,8 @@ export const secretApprovalRequestServiceFactory = ({
     );
     const hasMinApproval =
       secretApprovalRequest.policy.approvals <=
-      secretApprovalRequest.policy.approvers.filter(
-        ({ userId: approverId }) => reviewers[approverId.toString()] === ApprovalStatus.APPROVED
+      secretApprovalRequest.policy.approvers.filter(({ userId: approverId }) =>
+        approverId ? reviewers[approverId.toString()] === ApprovalStatus.APPROVED : null
```
Comment (DanielHougaard) [inline backend/src/ee/services/secret-approval-request/secret-approval-request-service.ts:451]: Nit: Would be more clear if you change `null` to `false`

Comment (DanielHougaard) [inline backend/src/ee/services/secret-approval-request/secret-approval-request-service.ts:451]: Why are we doing approverId.toString() when we are already checking if it's defined, by using the ternary operator?

**Review by DanielHougaard — CHANGES_REQUESTED** [review]
Looks good! I left a few comments

**Review by DanielHougaard — APPROVED** [review]
lgtm, good work!

**Pull request merged**